internal/envoy/v3/listener_test.go

```diff
@@ -1005,15 +1004,186 @@ func TestTCPProxy(t *testing.T) {
 // TestBuilderValidation tests that validation checks that
 // DefaultFilters adds the required HTTP connection manager filters.
 func TestBuilderValidation(t *testing.T) {
-	if err := HTTPConnectionManagerBuilder().Validate(); err == nil {
-		t.Errorf("ConnectionManager with no filters passed validation")
-	}
 
-	if err := HTTPConnectionManagerBuilder().AddFilter(&http.HttpFilter{Name: "foo"}).Validate(); err == nil {
-		t.Errorf("ConnectionManager with no filters passed validation")
+	assert.Error(t, HTTPConnectionManagerBuilder().Validate(),
+		"ConnectionManager with no filters should not pass validation")
+
+	assert.Error(t, HTTPConnectionManagerBuilder().AddFilter(&http.HttpFilter{
+		Name: "foo",
+		ConfigType: &http.HttpFilter_TypedConfig{
+			TypedConfig: &any.Any{
+				TypeUrl: "foo",
+			},
+		},
+	}).Validate(),
+		"ConnectionManager with only non-router filter should not pass validation")
+
+	assert.NoError(t, HTTPConnectionManagerBuilder().DefaultFilters().Validate(),
+		"ConnectionManager with default filters failed validation")
+
+	badBuilder := HTTPConnectionManagerBuilder().DefaultFilters()
+	badBuilder.filters = append(badBuilder.filters, &http.HttpFilter{
+		Name: "foo",
+		ConfigType: &http.HttpFilter_TypedConfig{
+			TypedConfig: &any.Any{
+				TypeUrl: "foo",
+			},
+		},
+	})
+	assert.Errorf(t, badBuilder.Validate(), "Adding a filter after the Router filter should fail")
+}
+
+func TestAddFilter(t *testing.T) {
+
+	tests := map[string]struct {
+		builder *httpConnectionManagerBuilder
+		add     *http.HttpFilter
+		want    []*http.HttpFilter
+	}{
+		"Nil add to empty builder": {
+			builder: HTTPConnectionManagerBuilder(),
+			add:     nil,
+			want:    nil,
+		},
+		"Add a single router filter to empty builder": {
+			builder: HTTPConnectionManagerBuilder(),
+			add: &http.HttpFilter{
+				Name: "router",
+				ConfigType: &http.HttpFilter_TypedConfig{
+					TypedConfig: &any.Any{
+						TypeUrl: HTTPFilterRouter,
+					},
+				},
+			},
```

Comment:
Seems like we could make a helper:
```Go
func TypedConfigOf(msg proto.Message) *TypedConfig {
        return http.HttpFilter_TypedConfig {
                TypedConfig: protobuf.MustMarshalAny(msg),
        }
}
```

then 
```Go
&http.HttpFilter{Name: "grpcweb", ConfigType: TypedConfigOf(&GrpcWeb{}))
```

Reply:
This meant we had to swap to using a protobuf comparer for the tests, which I feel is a bigger change than this warrants. I'll leave it for now.